### PR TITLE
Handle unsigned types in sign analysis (C# and Java)

### PR DESCRIPTION
**Pull request by tamasvajk**

### CODE REVIEW

**Review by hvitved — APPROVED** [review]
LGTM

File: java/ql/src/semmle/code/java/dataflow/internal/rangeanalysis/SignAnalysisSpecific.qll [hunk 1]
```diff
@@ -53,6 +53,12 @@ private module Impl {
   private import SignAnalysisCommon
   private import SsaReadPositionCommon
 
+  class UnsignedNumericType extends NumericOrCharType {
```
Comment (aschackmull) [inline java/ql/src/semmle/code/java/dataflow/internal/rangeanalysis/SignAnalysisSpecific.qll:56]: This class is already defined in `Type.qll` with the name `CharacterType`, so `UnsignedNumericType` could just alias that.

**Comment by tamasvajk** [comment]
Java differences job: https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/936/

**Comment by tamasvajk** [comment]
> Java differences job: https://jenkins.internal.semmle.com/job/Changes/job/Java-Differences/936/

The java differences job shows some changes, but they seem independent of this change.

**Comment by aschackmull** [comment]
> The java differences job shows some changes, but they seem independent of this change.

No? They are indeed very dependent on this change - and the changes look good. Indeed these are FPs that are now fixed due to additional sign information on the `char` type.

**Comment by tamasvajk** [comment]
> > The java differences job shows some changes, but they seem independent of this change.
> 
> No? They are indeed very dependent on this change - and the changes look good. Indeed these are FPs that are now fixed due to additional sign information on the `char` type.

Ahh, you're right. `tainted-arithmetic` and `uncontrolled-arithmetic` are actually checks that are highly dependent on this change. And `similar-file` is just adding noise in the results.

**Review by hvitved — APPROVED** [review]

**Comment by tamasvajk** [comment]
@aschackmull Does this PR work for you too? Can I merge it?

**Comment by aschackmull** [comment]
Yep, looks good!

**Pull request merged**